src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncPivotBlockSelector.scala

```diff
@@ -0,0 +1,144 @@
+        startRetryInterval
+      )
+      scheduleRetry(startRetryInterval)
+      context become idle
+    }
+  }
+
+  def tryChooseTargetBlock(peersWithBestBlockNumbers: Map[Peer, BigInt]): Unit = {
+    val peersSortedByBestNumber = peersWithBestBlockNumbers.toList.sortBy(-_._2)
+    val bestPeerBestBlockNumber = peersSortedByBestNumber.head._2
+    val expectedPivotBlock = (bestPeerBestBlockNumber - syncConfig.pivotBlockOffset).max(0)
+    val peersToAsk = peersSortedByBestNumber.takeWhile(_._2 >= expectedPivotBlock).map(_._1)
+
+    log.info(
+      "Trying to choose fast sync pivot block using {} peers. The best block is {}. Ask {} peers for block nr {}",
+      peersWithBestBlockNumbers.size,
+      bestPeerBestBlockNumber,
+      peersToAsk.size,
+      expectedPivotBlock
+    )
+
+    peersToAsk.foreach { peer =>
+      peerEventBus ! Subscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peer.id)))
+      etcPeerManager ! EtcPeerManagerActor.SendMessage(
+        GetBlockHeaders(Left(expectedPivotBlock), 1, 0, reverse = false),
+        peer.id
+      )
+    }
+
+    val timeout = scheduler.scheduleOnce(peerResponseTimeout, self, PivotBlockTimeout)
+    context become waitingForPivotBlock(peersToAsk.map(_.id).toSet, expectedPivotBlock, timeout, Map.empty)
+  }
+
+  def waitingForPivotBlock(
+      peersToAsk: Set[PeerId],
+      targetBlockNumber: BigInt,
+      timeout: Cancellable,
+      headers: Map[BlockHeader, Int]
+  ): Receive =
+    handleCommonMessages orElse {
+      case MessageFromPeer(blockHeaders: BlockHeaders, peerId) =>
+        peerEventBus ! Unsubscribe(MessageClassifier(Set(BlockHeaders.code), PeerSelector.WithId(peerId)))
+        val updatedPeersToAsk = peersToAsk - peerId
+        val targetBlockHeaderOpt = blockHeaders.headers.find(header => header.number == targetBlockNumber)
+
+        targetBlockHeaderOpt match {
+          case Some(targetBlockHeader) =>
+            log.info("Received vote for {} from {}", targetBlockHeader.hashAsHexString, peerId.value)
+            val newValue = headers.find(_._1 == targetBlockHeader).map(_._2 + 1).getOrElse(1)
+            val updatedHeaders = headers.updated(targetBlockHeader, newValue)
+            val (mostPopularBlockHeader, votes) = updatedHeaders.maxBy(_._2)
+            if (votes >= minPeersToChoosePivotBlock) {
+              timeout.cancel()
+              sendResponseAndCleanup(mostPopularBlockHeader)
+            } else if (updatedPeersToAsk.size + votes < minPeersToChoosePivotBlock) {
+              timeout.cancel()
+              peerEventBus ! Unsubscribe()
+              log.info("Not enough votes for pivot block. Retrying in {}", startRetryInterval)
+              scheduleRetry(startRetryInterval)
+              context become idle
+            } else {
+              context become waitingForPivotBlock(updatedPeersToAsk, targetBlockNumber, timeout, updatedHeaders)
+            }
+          case None =>
```

Comment:
Just wondering, if one of the peers did not respond or responded badly, maybe we should restart whole process if we won't have enough votes ?